### PR TITLE
Calculate the HFB weights

### PR DESCRIPTION
**Pull request by james-s-willis**

### CODE REVIEW

**Comment by james-s-willis** [comment]
@jrs65 I've also tidied the code up to use the `HFBFrameview`.

File: lib/stages/HFBAccumulate.cpp
Comment (james-s-willis) [inline lib/stages/HFBAccumulate.cpp:122]: @jrs65 I had to use `in_frame.hfb.data()[i]` instead of just `in_frame.hfb[i]` as it seg faults otherwise. Any idea what could cause this?

File: lib/stages/HFBAccumulate.cpp
Comment (jrs65) [inline lib/stages/HFBAccumulate.cpp:245]: If I had to pick one place that I thought was responsible for the incorrect weights, it is here. Both the array `input` and vector `hfb_even` are `int32_t` typed even though the underlying data is floating point. This subtraction operation will do an integer subtraction, which is then converted into a float. That will give the wrong answer, you need to do a floating point subtraction initially.

Ultimately the issue here is that none of the data is `int32_t`, if you type all the arrays and vectors appropriately this issue should disappear.

**Review by jrs65 — CHANGES_REQUESTED** [review]

File: lib/stages/HFBAccumulate.cpp
Comment (james-s-willis) [reply lib/stages/HFBAccumulate.cpp:199]: Ah of course, I forgot the input frames are from the GPU. And I thought there was a reason you took the data as `int32_t` in `visAccumulate`, but now I know to just take the data as a `float` in `HFBAccumulate`.

Comment (andrerenard) [reply lib/stages/HFBAccumulate.cpp:199]: The data from the N2 kernel is an `int32_t`, but in the case of the beams they are `float`

Comment (james-s-willis) [reply lib/stages/HFBAccumulate.cpp]: Ah that's why, thanks for the clarification.

**Comment by andrerenard** [comment]
What's left to do on this before we can merge?

**Comment by jrs65** [comment]
I think I'm meant to do a last review. I'll do that now.

File: lib/stages/HFBAccumulate.hpp
Comment (jrs65) [inline lib/stages/HFBAccumulate.hpp:82]: Given that we probably won't ever have gated HFB acq's you don't really need the internalState. You could just put all of it's state on the main class. Not one to do now, but probably worth noting for future cleanups.

File: lib/stages/HFBAccumulate.cpp
Comment (jrs65) [inline lib/stages/HFBAccumulate.cpp:296]: Note that after Rick's 2nd stage RFI changes the incoming data will have a dataset_id which needs to be used. I couldn't put a comment in the relevant in the place, but this means that the code which generates the dataset_id needs to be deferred until it receives data and can actually use that as the base for the new dataset id.

**Review by jrs65 — CHANGES_REQUESTED** [review]
Nearly there. A few small cleanups to request, and one slightly bigger one to make sure your changes work with Rick's new changes.

**Comment by andrerenard** [comment]
I think the conflicts with `develop` are blocking this from running the CI.   Can you update with `develop` and see if that helps?

**Comment by andrerenard** [comment]
It looks like this is failing on the chimeMetadata includes.  We moved it from `.h` to `.hpp`

File: lib/stages/HFBAccumulate.cpp
Comment (james-s-willis) [reply lib/stages/HFBAccumulate.cpp]: I think I have addressed this in my latest commit.

File: lib/stages/FreqSubset.hpp
Comment (james-s-willis) [reply lib/stages/FreqSubset.hpp:24]: Can this be deferred?

Comment (jrs65) [reply lib/stages/FreqSubset.hpp:24]: Definitely.

File: lib/stages/HFBAccumulate.hpp
Comment (jrs65) [inline lib/stages/HFBAccumulate.hpp:37]: The docstring here doesn't list the cls buffer as an input.

**Review by jrs65 — APPROVED** [review]
@james-s-willis I have one comment about a missing docstring. Otherwise this is done. Nice work! 

I've clicked approve, please do fix the docstring, but no need to come back to me for another review afterwards.

File: lib/stages/HFBAccumulate.hpp
Comment (james-s-willis) [reply lib/stages/HFBAccumulate.hpp:36]: Thanks Richard! :+1:

**Pull request merged**
